nixos/systemd-networkd: add PREF64 related options

jmbaur opened the pull request:
## Description of changes

Tested to work in the legacy nested `*Config` format as well as the newer flattened format.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jmbaur committed:
nixos/systemd-networkd: add PREF64 related options

flokli commented:
Is there a chance we can actually add a test for this?

jmbaur commented:
> Is there a chance we can actually add a test for this?

Sure thing! A functional test or a test that the config is what we expect?

mweinelt commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -2013,6 +2023,16 @@ let
 
+    ipv6PREF64Prefixes = mkOption {
+      default = [];
+      example = [ { Prefix = "2001:db8::/96"; } ];
```
```suggestion
      example = [ { Prefix = "64:ff9b::/96"; } ];
```

https://www.rfc-editor.org/rfc/rfc6052

jmbaur replied:
I don't mind either way, but we have a few other places in nixpkgs where NAT64-related things are being done, and we use the rfc3849 documentation prefix (albeit with varying prefix sizes) for those. I guess it's a matter of how the examples are interpreted, as an example of how a valid config looks, or as an example of a config that would be used in the real world. For consistency, I think keeping it as is is preferred.

mweinelt commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -2013,6 +2023,16 @@ let
 
+    ipv6PREF64Prefixes = mkOption {
+      default = [];
+      example = [ { Prefix = "64:ffb9::/96"; } ];
```
ff9b, not ffb9

jmbaur replied:
Thanks!

jmbaur committed:
nixos/clatd: make client fetch PLAT prefix dynamically

We can expose the PLAT prefix to the client via DNS64 so clatd is able
to determine the prefix dynamically. We can also test that some
systemd-networkd PREF64 settings work as expected when exposed on the
router.

jmbaur commented:
@flokli So we do have a clatd test! I added a commit that augments that test to ensure the networkd options work as expected.

jmbaur commented:
@ofborg test clatd

flokli commented on nixos/tests/clatd.nix:
```diff
@@ -102,6 +103,37 @@ import ./make-test-python.nix ({ pkgs, lib, ... }:
+        };
+      };
+
+      services.resolved.extraConfig = ''
```
So resolved does not have any support for this yet? Is there an upstream issue we can link to?

I assume once that's supported, we would not need to do any custom DNS setup here?

jmbaur replied:
This config is just set as to not conflict with the coredns instance running. clatd will fetch the PLAT prefix via DNS64, which is served up by coredns in the configuration I made.

jmbaur replied:
Ah I see what you're saying, does resolved have support for serving up DNS64 synthesized records. AFAIK no it does not, and I don't see any issue upstream.

flokli replied:
Yes, exactly. Would you feel like opening one?

- This feels like a logical successor after the pref64 stuff, so it would make sense to signal interest in it
- It would allow others looking at the code and module options to understand why we went with the coredns route 
- it makes it easy to know when we can drop coredns from here again

jmbaur replied:
I'm actually not super sure that adding a DNS64 component to systemd-resolved would be a good idea. If a client running systemd-networkd is already able to obtain the PREF64 prefix from the router advertisements, what good is it for systemd-resolved to serve up modified DNS responses to applications when the machine should be smart enough to setup a clat interface so the unmodified DNS A RRs can be used directly. I do get that we are kinda in a weird place right now where systemd-networkd exposes this information about the PLAT prefix existing, but doesn't (yet) take any responsibility in setting up a clat interface, and at the same time no software that I'm aware yet reads this exposed information, so software such as clatd is left to do "discovery" of the presence of a NAT64 machine by virtue of it receiving DNS64 responses. I think the logical move forward is for systemd-networkd to add proper 464xlat support and/or for more software to start reading from the info exposed by networkd.

In regards to using coredns for this test, I'm happy to use another DNS server if you think that would be best, I am just personally familiar with the coredns config and with the config being quite terse for this use case.

jmbaur replied:
Relevant systemd issue: https://github.com/systemd/systemd/issues/23674

flokli replied:
Ok yeah, fair points. I guess we can generally follow this as a tracking issue. Thanks for the digging!

flokli reviewed (approved):
LGTM, but I'd like to give @hax404 a chance to also review the test additions.

mweinelt commented:
I think it is reasonable to backport this change, as it introduces the declarative interface for functionality present in systemd 255.

hax404 reviewed (approved):
Very nice work!

flokli commented:
@mweinelt please do the backport manually.
